pkg/kubelet/cm/memorymanager/policy_single_numa.go

```diff
@@ -17,50 +17,340 @@ limitations under the License.
+	"k8s.io/klog"
+	corehelper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	v1qos "k8s.io/kubernetes/pkg/apis/core/v1/helper/qos"
 	"k8s.io/kubernetes/pkg/kubelet/cm/memorymanager/state"
 	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
+	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager/bitmask"
 )
 
 const policyTypeSingleNUMA policyType = "single-numa"
 
 // TODO: need to implement all methods
 
+type reservedMemory map[int]map[v1.ResourceName]uint64
+
 // SingleNUMAPolicy is implementation of the policy interface for the single NUMA policy
 // TODO: need to re-evaluate what field we really need
 type singleNUMAPolicy struct {
+	// machineInfo contains machine memory related information
+	machineInfo *cadvisorapi.MachineInfo
+	// reserved contains memory that reserved for kube
+	systemReserved reservedMemory
 	// topology manager reference to get container Topology affinity
 	affinity topologymanager.Store
 }
 
 var _ Policy = &singleNUMAPolicy{}
 
 // NewPolicySingleNUMA returns new single NUMA policy instance
-func NewPolicySingleNUMA() Policy {
-	return &singleNUMAPolicy{}
+func NewPolicySingleNUMA(machineInfo *cadvisorapi.MachineInfo, reserved reservedMemory, affinity topologymanager.Store) Policy {
+	// TODO: check if we have enough reserved memory for the system
+	//for _, node := range reserved {
+	//	if node[v1.ResourceMemory] == 0 {
+	//
+	//	}
+	//}
+	return &singleNUMAPolicy{
+		machineInfo:    machineInfo,
+		systemReserved: reserved,
+		affinity:       affinity,
+	}
 }
 
 func (p *singleNUMAPolicy) Name() string {
 	return string(policyTypeSingleNUMA)
 }
 
 func (p *singleNUMAPolicy) Start(s state.State) error {
+	if err := p.validateState(s); err != nil {
+		klog.Errorf("[memorymanager] Invalid state: %v, please drain node and remove policy state file", err)
+		return err
+	}
 	return nil
 }
 
 // Allocate call is idempotent
 func (p *singleNUMAPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Container) error {
+	// allocate the memory only for guaranteed pods
+	if v1qos.GetPodQOS(pod) != v1.PodQOSGuaranteed {
+		return nil
+	}
+
+	klog.Infof("[memorymanager] Allocate (pod: %s, container: %s)", pod.Name, container.Name)
+	if blocks := s.GetMemoryBlocks(string(pod.UID), container.Name); blocks != nil {
+		klog.Infof("[memorymanager] Container already present in state, skipping (pod: %s, container: %s)", pod.Name, container.Name)
+		return nil
+	}
+
+	// Call Topology Manager to get the aligned affinity across all hint providers.
+	hint := p.affinity.GetAffinity(string(pod.UID), container.Name)
+	klog.Infof("[memorymanager] Pod %v, Container %v Topology Affinity is: %v", pod.UID, container.Name, hint)
+
+	if !hint.Preferred {
+		return fmt.Errorf("[memorymanager] failed to get preferred NUMA affinity")
+	}
+
+	var affinityBits []int
+	if hint.NUMANodeAffinity == nil {
+		if len(p.machineInfo.Topology) > 1 {
+			return fmt.Errorf("[memorymanager] machine has more than one NUMA node, but NUMA affinity is empty")
```

Comment:
the same as above, validation for alignment of memory and topology manager policies will solve this issue


> * <sth, true> => choose given NUMA node & pin memory
> * <nil, true> => choose a NUMA node & pin memory anyway

possible only in case when best hint equal to the default NUMA nodes set, and it possible only for single NUMA nodes machines

> * <nil, false> => choose a NUMA node & pin memory anyway

Not possible at all, under the single NUMA node topology manager policy, all hints should be preferred, otherwise it will be filtered
https://github.com/kubernetes/kubernetes/blob/4c81edbb41f3f1a7e2d20b8ad63b6a2701421214/pkg/kubelet/cm/topologymanager/policy_single_numa_node.go#L51

Reply:
For the case`<nil, false>`,
I'm talking about the case when TM is disabled or is configured with None policy.
In those case, MM.Allocate() will be called without calling GetTopologyHint() and  TM.Policy.Merge().

See here for when TM is disabled
https://github.com/mgjeong/kubernetes/blob/209adb3614f1cc994314e08bbdb82e5fbaf6babd/pkg/kubelet/cm/container_manager_linux.go#L752

See here for when TM is configured with NONE policy.
https://github.com/mgjeong/kubernetes/blob/5bf4a4ca2fe91d279ccb316afcfa916172e52f4a/pkg/kubelet/cm/topologymanager/topology_manager.go#L238

Reply:
yep, but if we will have validation on the kubelet start, something like:
```
if mm.enabled && mm.policy == "single-numa" && (tm.disabled || tm.policy != "single-numa-node") {
    // drop the error and do not start the kubelet at all
}
```

Reply:
and IMHO it better to have a clear error message under the kubelet, instead of trying to understand why container allocation is failing